adr/ADR-40.md

```diff
@@ -52,209 +61,426 @@ This method is available in all NATS Server versions.
+other servers in the cluster of which it is aware. This can be very helpful for
+discoverability and removes the need for the client to pass all servers in
+`connect`, but it also may be unwanted if, for example, some servers URLs are
+unreachable for a given client.
 
 #### Retain servers order
 
-**default: false**
-By default, if many server addresses are passed in the connect string or array, the client will try to connect to them in random order.
-This helps healthy connection distribution, but if in a specific case list should be treated as a preference list,
-randomization may be turned off.
+**default: false** By default, if many server addresses are passed in the
+connect string or array, the client will try to connect to them in random order.
+This helps healthy connection distribution, but if in a specific case list
+should be treated as a preference list, randomization may be turned off.
 
-This function can be expressed "enable retaining order" or "disable randomization" depending on what is more idiomatic in given language.
+This function can be expressed "enable retaining order" or "disable
+randomization" depending on what is more idiomatic in given language.
 
 ### Protocol Commands and Grammar
 
 #### INFO
+
 [LINK][LINK]
 
-Send by the Server before or after establishing TLS, depending of flow used.
-It contains information about the Server, the nonce, and other server URLs to which the client can connect.
+Send by the Server before or after establishing TLS, depending of flow used. It
+contains information about the Server, the nonce, and other server URLs to which
+the client can connect.
 
 #### CONNECT
-[CONNECT][CONNECT]
 
-Send by the client in response to INFO.
-Contains information about client, including optional signature, client version and connection options.
+[CONNECT][CONNECT]
 
+Send by the client in response to INFO. Contains information about client,
+including optional signature, client version and connection options.
 
 #### Ping Pong
-This is a mechanism to detect broken connections that may not be reported by the network connection in a given language.
 
-If the Server sends `PING`, the client should answer with `PONG`.
-If the Client sends `PING`, the Server should answer with `PONG`.
+This is a mechanism to detect broken connections that may not be reported by the
+network connection in a given language.
 
-If two (configurable) consecutive `PONGs are missed, the client should treat the connection as broken, and it should start reconnect attempts.
+If the Server sends `PING`, the client should answer with `PONG`. If the Client
+sends `PING`, the Server should answer with `PONG`.
 
-The default interval for PING is 2 minutes.
+If two (configurable) consecutive `PONGs are missed, the client should treat the
+connection as broken, and it should start reconnect attempts.
 
+The default interval for PING is 2 minutes.
 
 ### Error Handling (TODO)
```

Comment:
Missing Stale Connection, client should reconnect after getting this error:

https://github.com/nats-io/nats.go/blob/main/nats.go#L3655-L3665

Reply:
Roughly in the Go client, it reconnects when receiving either one of these:

- stale connection
- max connection exceeded

Does not reconnect and just triggers async callbacks:

- permissions and auth errors

Then for any other -ERR it will close the connection and not reconnect again:

https://github.com/nats-io/nats.go/blob/main/nats.go#L3647-L3675